lib/upgrade/upgrade_compat.go

```diff
@@ -0,0 +1,54 @@
+// Copyright (C) 2024 The Syncthing Authors.
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+package upgrade
+
+import (
+	"fmt"
+	"runtime"
+	"strings"
+
+	"github.com/shirou/gopsutil/v4/host"
+)
+
+// verifyRuntimeRequirements returns an error if the OS's version (aka
+// kernel version) is incompatible with the version of go used to build the
+// release. If there is no requirement for the current OS and architecture,
+// no error is returned.
+func verifyRuntimeRequirements(runtimeReqs RuntimeReqs) error {
+	currentKernel, err := host.KernelVersion()
+	if err != nil {
+		return err
+	}
+	currentKernel = normalizeKernelVersion(currentKernel)
+
+	for hostArch, minKernel := range runtimeReqs.Requirements {
+		host, arch, found := strings.Cut(hostArch, "/")
+		if host != runtime.GOOS {
+			continue
+		}
+		if found && arch != runtime.GOARCH {
+			continue
+		}
+		if CompareVersions(minKernel, currentKernel) > Equal {
+			return fmt.Errorf("runtime %v requires OS version %v but this system is version %v", runtimeReqs.Runtime, minKernel, currentKernel)
+		}
+	}
```

Comment:
Do we think yolo upgrading because we can't find the goos/arch combo is right?

Reply:
I think we have to, otherwise we need to cover all possible OSes. And we currently don't and I doubt we even could if we wanted to, there's always going to be something esoteric we'll miss.